Associated type resilience

DougGregor opened the pull request:
Make associated types resilient against reordering, without depending on the "sorting" hack that we previously had in place. This is setting the stage to be able to resiliently add new, defaulted associated types without breaking backward compatibility.

DougGregor committed:
[ABI] Emit associated type descriptors referencing each requirement.

Emit associated type descriptors (as aliases) to reference each associated
type requirement within a resilient protocol.

DougGregor committed:
[ABI] Add protocol requirements base descriptor.

Introduce an alias that refers one element prior to the start of a
protocol descriptor’s protocol requirements. This can be subtracted from
an associated type descriptor address to determine the offset of the
associated type accessor within a corresponding witness table. The code
generation for the latter is not yet implemented.

slavapestov commented on docs/ABI/Mangling.rst:
```diff
@@ -138,6 +138,10 @@ types where the metadata itself has unknown layout.)
   global ::= type generic-signature 'Th' // key path hasher
 
+  global ::= protocol 'TL'               // protocol requirements base descriptor
+  global ::= assoc-type-name 'Tl'        // associated type descriptor
```
Why not use the same Tq suffix we use for method descriptors? There won't be a conflict and you'll be able to reuse l for other things

slavapestov replied:
What about conformances? Are we just going to say you can't add new conformance requirements after the fact, and existing conformances have a canonical ordering?

DougGregor replied:
I didn't re-use `Tq` because IIUC the mangling grammar is usually not resolved this way: the suffix always (?) tells you what grammar to expect before it. Now, I am stealing several of the thunk mangling  (for the requirements base, associated type, and possibly associated conformance descriptors), so perhaps I should move out to three-letter discriminators because there aren't that many of these.

Associated conformances are interesting. We need to be able to add new conformances for new defaulted associated types, but we can't add any to existing associated types. Introducing associated conformance descriptors would be the simplest way of ensuring that the addition of new requirements doesn't affect existing clients, so I was planning to investigate that next.

slavapestov replied:
If new conformances are only ever added to new associated types, you might be able to get away with not giving them their own descriptors no?

DougGregor replied:
If we can find a way to consistently put them at the end, we wouldn't need descriptors for associated conformances. But that gets back to the "ABI history database" issue.

slavapestov commented on lib/IRGen/IRGenMangler.h:
```diff
@@ -16,6 +16,7 @@
 #include "IRGenModule.h"
 #include "swift/AST/ASTMangler.h"
 #include "swift/IRGen/ValueWitness.h"
+#include "llvm/Support/SaveAndRestore.h"
```
Is this needed?

DougGregor replied:
Yes, it's used so we don't drop the protocol name in the mangling.

DougGregor replied:
Hmm. I could mangle this as <protocol> <assoc-type-name> and let it drop the associated type names from the rest of the mangling.

DougGregor committed:
[IRGen] witness_method should always refer to a slot in the witness table.

DougGregor committed:
[IRGen] Implement resilient associated type witness access pattern.

When accessing the associated type witness metadata for a resilient
protocol, compute the index based on the difference between the
associated type’s descriptor and the protocol requirement base descriptor
to determine the offset into the witness table.

DougGregor commented:
@swift-ci please smoke test

slavapestov commented:
Can you also add a test to validation-test/Evolution/?

DougGregor committed:
[Runtime] Fix _ObjectiveCBridgeableWitnessTable::ObjectiveCType

The signature of this function pointer was incorrect, because it had not
yet been updated for the metadata request/response architecture. We
were getting lucky (!).

DougGregor committed:
[ABI] Emit associated type witnesses resiliently.

Emit associated type witnesses into resilient conformance tables, so they
can be re-ordered within the protocol without breaking clients. This will
(eventually) permit adding new, defaulted associated types to protocols
resiliently.

DougGregor commented:
> Can you also add a test to validation-test/Evolution/?

It turns out that we're already testing for reordering of associated types in https://github.com/apple/swift/blob/master/validation-test/Evolution/Inputs/protocol_reorder_requirements.swift#L29-L53; it worked before my pull request because we sorted associated types. Now we don't sort them, so it's working "for real'.

I have not implementing resiliently adding an associated type (yet). That will need a new test.

DougGregor committed:
[IRGen] Simplify nondependent associated type witness metadata accessors.

When we’re creating an associated type witness metadata accessor for
resilience reasons, but the associated type witness doesn’t involve any
type parameters, directly form the type metadata reference (and don’t
cache it).

While here… update all of the IRGen/SILGen test cases perturbed by the
introduction of resilient associated type access patterns.

DougGregor committed:
[ABI] Use associated type descriptors for generic parameter references.

Generic parameter references, which occur in generic requirement
metadata, were hardcoding associated type indices. Instead, use
relative references to associated type descriptors and perform the
index calculation at runtime.

Associated types can now be reordered resiliently (without relying on 
sorting), which is the first main step toward rdar://problem/44167982.

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please benchmark

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
Small code size regression, as expected.

DougGregor committed:
Adjust test for Linux

DougGregor committed:
[IRGen] Fix associated type resilience for 32-bit platforms.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
@swift-ci please test macOS

DougGregor committed:
Another test fix for Linux

DougGregor commented:
@swift-ci please test
